### PR TITLE
Add GPU detection and auto-assign GPUs

### PR DESCRIPTION
**Pull request by shuds13**
*Re-opened after merge from develop (was #980) - now going straight to develop, superseding #928

----

Extension to gpu_detect branch. Executor options to set CPUs and GPUs to slots using either the method supplied in configuration or determined by the detected environment. This simplifies sim functions and makes them more portable.

There are different ways this could be specified in the arguments to submit. Current approach uses`auto_assign_gpus` and optionally `match_procs_to_gpus`, where `auto_assign_gpus` will use all GPUs assigned to current worker. Method for setting GPUs determined by either MPI runner, specified platform (via env variable LIBE_PLATFORM), or default.

### Checklists

Before merging:

- [x] Check all prints removed
- [x] Check TODO comments (and any others no longer relevant) removed
- [x] Docstrings correct for new functions
- [x] type-hints correct for new functions

1. Initial single node testing (using CUDA var resources reg. test and forces_gpu):

- [x] Laptop (inc dry_run mock mpi_runners)
- [x] Spock (works with detection - uses srun options). Can also work with ROCR_VISIBLE_DEVICES.
- [x] Crusher (works with detection - issue across GCDs not caused by this PR - known issue)
- [x] Summit (detection works)
- [x] Perlmutter (works with srun given - or system env var - also now works with default mpirun/CVD!)
- [x] Polaris
- [x] Frontier (works with detection, even if don't provide platform).
- [x] Swing 
- [ ] ~~Sunspot - currently need to set GPU/worker (and MPI rank) affinity via a bash script. Sep. branch for this.~~

2. Advanced testing: multiple nodes / mpi4py comms / awkward partitions.
- [x] Frontier
- [x] Summit
- [x] Perlmutter
- [x] Polaris

### CODE REVIEW

File: docs/data_structures/libE_specs.rst [hunk 1]
```diff
@@ -148,6 +148,17 @@ upon passing into :meth:`libE()<libensemble.libE.libE>`.
...
             Default: False
+        "platform" [str]:
+            Name of a known platform (see "Known Platforms List" below)
+            E.g., libE_specs["platform"] = "perlmutter_g"
+            Note: the environment variable LIBE_PLATFORM is an alternative way of setting.
+            See also option "platform_specs".
+        "platform_specs" [Platform|dict]:
+            A Platform object (or dictionary) specifying settings for a platform.
+            For list of Platform fields see "Platform Fields" below.
+            Any fields not given, wil be auto-detected by libEnsemble.
```
Comment (jlnav) [inline docs/data_structures/libE_specs.rst:159]: typo

File: docs/platforms/frontier.rst [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Frontier_ is an HPE Cray EX exascale system located at Oak Ridge
+Leadership Computing Facility (OLCF).
+
+Each Frontier compute node contains a 64-core AMD EPYC and 4 AMD MI250X GPUs
+(8 logical GPUs).
+
+Frontier uses the SLURM scheduler to submit jobs from login nodes to run on the
+compute nodes.
+
+Installation libEnsemble
```
Comment (jlnav) [inline docs/platforms/frontier.rst:14]: "libEnsemble Installation" or "Installing libEnsemble"

File: docs/platforms/frontier.rst [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    module load cray-python
+
+You may wish to create a virtual environment to install packages in (see python_on_frontier_).
+
+.. container:: toggle
+
+   .. container:: header
+
+      Example of using virtual environment
```
Comment (jlnav) [inline docs/platforms/frontier.rst:27]: On develop we have a better drop-down solution now:

```
.. dropdown:: Example of using virtual environment
```

can replace 23-27

File: docs/platforms/perlmutter.rst [hunk 4]
```diff
@@ -12,7 +13,7 @@ Configuring Python and Installation
 
 Begin by loading the ``python`` module. The following modules are recommended::
 
-    module load PrgEnv-gnu cudatoolkit python
+    module python
```
Comment (jlnav) [inline docs/platforms/perlmutter.rst:16]: module "load" python

File: docs/platforms/perlmutter.rst [hunk 5]
```diff
@@ -49,130 +50,130 @@ by one of the following ways.
...
-                        extra_args="--gpus-per-task=1"
-                        )
+.. note::
 
-If running using :doc:`variable resource workers<../resource_manager/overview>`,
-between one and four-way MPI runs may be issued by any of the workers (with each
-MPI task using a GPU). libEnsemble's resource manager automatically disables workers
-whose resources are being used by another worker.
+    The video is made from libEnsemble version 0.9.3, where some adjustments needed to be
+    made to the scripts to run on Perlmutter. These are no longer be necessary. libEnsemble
```
Comment (jlnav) [inline docs/platforms/perlmutter.rst:103]: "These are no longer necessary"

File: docs/platforms/perlmutter.rst [hunk 6]
```diff
@@ -49,130 +50,130 @@ by one of the following ways.
...
+
+Running libEnsemble with local comms is usually sufficient on Perlmutter. However, if you need
+to use ``mpi4py``, you should install and run as follows::
+
+    module load PrgEnv-gnu cudatoolkit
+    MPICC="cc -target-accel=nvidia80 -shared" pip install --force --no-cache-dir --no-binary=mpi4py mpi4py
+
+This line will build ``mpi4py`` on top of a CUDA-aware Cray MPICH.
+
+To run using 4 workes (one manager)::
```
Comment (jlnav) [inline docs/platforms/perlmutter.rst:121]: typo

File: docs/platforms/perlmutter.rst [hunk 7]
```diff
@@ -49,130 +50,130 @@ by one of the following ways.
...
+in the `Python on Perlmutter`_ documentation.
+
 Perlmutter FAQ
 --------------
 
-**srun: Job \*\*\*\*\*\* step creation temporarily disabled, retrying (Requested nodes are busy)**
+Some FAQs specific to Perlmutter. See more on the :doc:`FAQ<../FAQ>` page.
 
-You may also see: ``srun: Job ****** step creation still disabled, retrying (Requested nodes are busy)``
+.. container:: toggle
```
Comment (jlnav) [inline docs/platforms/perlmutter.rst:134]: Same as before, new dropdown method

**Review by jlnav — APPROVED** [review]

**Review by jmlarson1 — APPROVED** [review]

**Pull request merged**